### ISSUE TITLE
The `account.signed` flow event can't be used to signal completion of a flow

### ISSUE THREAD

**Issue by philbooth** [mozilla/fxa-auth-server#1457]
I noticed this when digging into the data for the sign-in funnels.

Because device registration frequently occurs before a sign-in is confirmed or sign-up is verified, we decided to use the first `account.signed` event as the signal for completion of a flow. The thinking was this would indicate that the device is now _really_ registered, but that's not the case because the content server also hits that endpoint when creating an OAuth token for use with the profile server. So if device registration fails, or is never requested at all, flows will still show up as completed.

[Here is a query showing these rogue "completed" flows](https://sql.telemetry.mozilla.org/queries/1210/source).

One way to fix it would be to constrain the `account.signed` flow event if `device.created` has not yet been emitted. Or maybe we can reliably detect that the requester is Sync before emitting the event. Both of these approaches are problematic in that they will not take effect for historical data. But flow events are so new that maybe we don't care about that yet.

Alternatively we could fix it at the SQL level instead, and ignore `account.signed` unless it joins to `device.created`. In the cases where it doesn't, `device.created` would be the correct signal to use for a completed flow. This will lead to messier queries but has the advantage of working for historical data.

Having written that down, I think I prefer the server-side fix.

@rfk, thoughts?

**Comment by philbooth** [comment]
> One way to fix it would be to constrain the `account.signed` flow event if `device.created` has not yet been emitted.

Fwiw I've started down this road, leaning on our old friend memcached to store the `device.created` state.

**Comment by rfk** [comment]
We do have the power the tell the differences between cert-signs from content-server and cert-signs from the browser, IIRC we do it in the code for registering placeholder devices.  Can we use a similar trick in the metrics?

**Comment by rfk** [comment]
Hrm, so I think I see - on the server we use the `service` parameter to distinguish between cert-signs from content-server and from the device.  But if they're both part of the same flow, we only have a single value for the `service` parameter.

ISTM that perhaps the cert-sign event from content-server should, somehow, just not be included in the flow metrics.  Perhaps we can detect on the server and not emit the flow event in this case?

**Comment by rfk** [comment]
@philbooth in the linked graph, I notice that the vast majority of the records are on fennec, and some are on very old versions of desktop firefox.  I wonder if this is something to do with placeholder device records not properly emitting the `device.created` event?

**Comment by rfk** [comment]
> Both of these approaches are problematic in that they will not take effect for historical data.
> But flow events are so new that maybe we don't care about that yet.

I'd rather get it right going forward, especially since the data is so new.  Perhaps we can do some one-off post-processing to fix up the historical data in redshift.

**Comment by philbooth** [comment]
> Hrm, so I think I see - on the server we use the `service` parameter to distinguish between cert-signs from content-server and from the device.  But if they're both part of the same flow, we only have a single value for the `service` parameter.
>
> ISTM that perhaps the cert-sign event from content-server should, somehow, just not be included in the flow metrics.  Perhaps we can detect on the server and not emit the flow event in this case?

I like this plan, but we also need to consider what to do about non-Sync sessions.

Currently we emit flow events for all sessions, not just Sync. Many of those non-Sync sessions are currently accidentally showing up as completed because of this bug (so right result for the wrong reason). If we fix it in this way, we'll need to change the re:dash queries to use different completion signals for non-Sync flows.

Alternatively, maybe we don't want the ability to analyse non-Sync flows in the funnel? In which case we should consider not emitting flow events for them at all.

**Comment by philbooth** [comment]
I also discovered two other anomalies yesterday, which may be manifestations of this problem or may be something different (or maybe it's both of those).

The first one is unverified users who trigger the sign-up completion signal:

https://sql.telemetry.mozilla.org/queries/1205/source

Is this related to https://github.com/mozilla/fxa-content-server/issues/4152?

The second one is unconfirmed users who trigger the sign-in completion signal:

https://sql.telemetry.mozilla.org/queries/1207/source

Note that there are plenty of desktop users in there who I would expect to go through sign-in confirmation.

**Comment by rfk** [comment]
Hmm, I don't have a ready explanation for either of those, apart from the possibility that (a) we've missed logging the events some how, or (b) there's some weird security-related problem that lets these users skip the verification step...

**Comment by rfk** [comment]
I picked a flow-id from your "unconfirmed users who trigger the sign-in completion signal" dashboard and took a look at all their events:

```
select * from flow_events where flow_id='730eda3f8c1982d43887e962d8ce1fc8a042695cfa10ee89378c4e330e15c483';
      timestamp      | flow_time |                             flow_id                              |       type       
---------------------+-----------+------------------------------------------------------------------+------------------
 2016-09-02 01:20:27 |         0 | 730eda3f8c1982d43887e962d8ce1fc8a042695cfa10ee89378c4e330e15c483 | flow.begin
 2016-09-02 01:21:57 |     89843 | 730eda3f8c1982d43887e962d8ce1fc8a042695cfa10ee89378c4e330e15c483 | account.created
 2016-09-02 01:22:16 |    109124 | 730eda3f8c1982d43887e962d8ce1fc8a042695cfa10ee89378c4e330e15c483 | account.login
 2016-09-02 01:22:18 |    110548 | 730eda3f8c1982d43887e962d8ce1fc8a042695cfa10ee89378c4e330e15c483 | device.created
 2016-09-02 01:24:37 |    249492 | 730eda3f8c1982d43887e962d8ce1fc8a042695cfa10ee89378c4e330e15c483 | account.keyfetch
 2016-09-02 01:24:37 |    249692 | 730eda3f8c1982d43887e962d8ce1fc8a042695cfa10ee89378c4e330e15c483 | account.signed
```

Strangely, it had both an `account.created` and an `account.login` on the same flow_id.

**Comment by rfk** [comment]
(although they're not all like that)

**Comment by philbooth** [comment]
Temporary hack landed in #1462. Deliberately leaving this issue open for implementation of the proper fix.

**Comment by philbooth** [comment]
This was fixed by mozilla/fxa-auth-server#1515.
